File: DungeonDebt/Assets/Scripts/Data/HeroInstance.cs
using System;

public class HeroInstance
{
    public HeroInstance(HeroDefinition definition, int formationSlot)
    {
        Definition = definition;
        CurrentHealth = definition.BaseHealth;
        Attack = definition.BaseAttack;
        UpkeepThisRound = definition.BaseUpkeep;
        FormationSlot = formationSlot;
        InstanceId = Guid.NewGuid();
        Tier = HeroTier.Bronze;
        VeteranXp = 0;
        VeteranTier = 0;
    }

    public HeroDefinition Definition { get; set; }
    public int CurrentHealth { get; set; }
    public int Attack { get; set; }
    public int UpkeepThisRound { get; set; }
    public int FormationSlot { get; set; }
    public Guid InstanceId { get; set; }
    public HeroTier Tier { get; set; }
    public int VeteranXp { get; set; }
    public int VeteranTier { get; set; }
}

File: web/src/data/HeroInstance.js
// Ported from DungeonDebt/Assets/Scripts/Data/HeroInstance.cs
// Mutable party member built from an immutable HeroDefinition.
import { HeroTier } from "./enums.js";

export class HeroInstance {
  constructor(definition, formationSlot) {
    this.definition = definition;
    this.currentHealth = definition.baseHealth;
    this.attack = definition.baseAttack;
    this.upkeepThisRound = definition.baseUpkeep;
    this.formationSlot = formationSlot;
    this.instanceId = (globalThis.crypto && crypto.randomUUID)
      ? crypto.randomUUID()
      : `${Date.now()}-${Math.random().toString(16).slice(2)}`;
    this.tier = HeroTier.Bronze;
    this.veteranXp = 0;
    this.veteranTier = 0;
  }
}
